Add win_msg module

jhawkesworth opened the pull request:
##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - New Module Pull Request


##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->
win_msg

##### ANSIBLE VERSION
<!--- Paste verbatim output from “ansible --version” between quotes below -->
```
ansible 2.3.0 (win_msg_new ad7cb10fc6) last updated 2017/02/27 17:15:23 (GMT +100)
  config file =
  configured module search path = Default w/o overrides
```

##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->
In keeping with the current zeitgeist of reviving all things nineties https://www.theregister.co.uk/2017/02/14/hmd_nokia_nostalgia/, this module adds the ability to irritate windows users just like you used to using pop up messages.

No tests at this time, but easy enough to test using ad-hoc ansible.

Due to the curious and ancient nature of the msg.exe messages are sent in turn to logged on users, which can result in long run times if you set the 'wait' parameter to 'true'.  For this reason I have added module runtime info to the module return output.

<!---
If you are fixing an existing issue, please include "Fixes #nnn" in your
commit message and your description; but you should still explain what
the change does.
-->

This change fixes nothing.

<!-- Paste verbatim command output below, e.g. before and after your change -->
```

```

Commit:
new implementation of win_msg.  TODO tests

jhawkesworth commented:
@dagwieers - please review if you have a moment.
@jborean93 - I tweaked this a little from the old version in modules-extra which you reviewed.  Would you mind having another go with this version?

dagwieers commented on lib/ansible/modules/windows/win_msg.ps1:
```diff
@@ -0,0 +1,64 @@
+
+$stopwatch = [system.diagnostics.stopwatch]::startNew()
+
+$params = Parse-Args $args
```
Enable check-mode support !

dagwieers commented on lib/ansible/modules/windows/win_msg.ps1:
```diff
@@ -0,0 +1,64 @@
+$stopwatch = [system.diagnostics.stopwatch]::startNew()
+
+$params = Parse-Args $args
+$result = New-Object PSObject
```
Use a normal hashtable.

dagwieers commented on lib/ansible/modules/windows/win_msg.ps1:
```diff
@@ -0,0 +1,64 @@
+$to = Get-Attr -obj $params -name to -default "*" -failifempty $False -resultobj $result
+$display_seconds = Get-Attr -obj $params -name display_seconds -default "10" -failifempty $False -resultobj $result
+$wait = Get-Attr -obj $params -name wait -default $False -failifempty $False -resultobj $result
+$msg = Get-Attr -obj $params -name msg -failifempty $True -resultobj $result
```
Use Get-AnsibleParam for all of these :-)

dagwieers commented on lib/ansible/modules/windows/win_msg.ps1:
```diff
@@ -0,0 +1,64 @@
+If($msg.Length -lt 1) {
+
+    Fail-Json $result "msg was empty, please supply message text"
+}
```
I would set a standard message by default. (Just like debug and fail also do)

dagwieers commented on lib/ansible/modules/windows/win_msg.py:
```diff
@@ -0,0 +1,95 @@
+  display_seconds:
+    description:
+      - How long to wait for receiver to acknowledge message.
+    required: false
```
Don't add `required: false`. It is implicit.

dagwieers commented on lib/ansible/modules/windows/win_msg.py:
```diff
@@ -0,0 +1,95 @@
+  to:
+    description:
+      - Who to send the message to. Can be a username, sessionname or sessionid.
+    required: false
```
Don't add `required: false`. It is implicit.

dagwieers commented on lib/ansible/modules/windows/win_msg.py:
```diff
@@ -0,0 +1,95 @@
+    description:
+      - Who to send the message to. Can be a username, sessionname or sessionid.
+    required: false
+    default: all logged in users on the windows target.
```
This doesn't seem right. Default states the default value, not some documentation string.

dagwieers commented on lib/ansible/modules/windows/win_msg.py:
```diff
@@ -0,0 +1,95 @@
+  wait:
+    description:
+      - Whether to wait for users to respond.  Module will only wait for the number of seconds specified in display_seconds or 10 seconds if not specified.  However, if I(wait) is true, the message is sent to each logged on user in turn, waiting for the user to either press 'ok' or for the timeout to elapse before moving on to the next user.
+    required: false
```
Don't add `required: false`. It is implicit.

dagwieers commented on lib/ansible/modules/windows/win_msg.py:
```diff
@@ -0,0 +1,95 @@
+    description:
+      - The text of the message to be displayed.
+    required: true
+    default: no default
```
Remove this line. If it is required, there is no default (not  "no default").
But I would add a default message.

Commit:
added check_mode.  Get-AnsibleParam for parameters.  Default msg is now 'Hello world!'. Use hash for $results.
Re-ordered some parameters alphabetically.
Documentation now in raw strings.  Removed spurious required: false.  Added metadata.

jhawkesworth commented:
Thanks for the review @dagwieers - I think I have fixed all the issues you mention now.

ready_for_review

Commit:
Keep pep8 checks happy

jhawkesworth commented:
Ok jumped the gun a bit there, latest push should keep pep8 checks happy.

ready_for_review

Commit:
Really keep pep8 checks happy this time

jhawkesworth commented:
.. only kidding.  3rd time lucky

ready_for_review

Commit:
update Changelog.md

jhawkesworth commented:
.. and this time I remembered to update the changelog

ready_for_review

dagwieers commented:
shipit

jhawkesworth commented:
bot_status

jhawkesworth commented:
Hmm, just one ship it short ...

jborean93 commented:
Sorry @jhawkesworth I will try and get to this later today for me but a lot of my spare time has been taken up by another project I am working on right now.

jborean93 commented on lib/ansible/modules/windows/win_msg.ps1:
```diff
@@ -0,0 +1,65 @@
+$msg_args = @($to, "/TIME:$display_seconds")
+
+If ($wait) 
+{
```
Are we able to keep the braces the same as the rest of the windows modules?

dagwieers replied:
@jborean93 There's no consensus, and there's no standard way of doing things in Powershell (cfr. PEP8 in Python). For that reason I removed all the formatting changes in my refactoring of most Windows modules on request of @nitzmahone 

I'd prefer a standard way of writing modules in absence of a higher authoritative source for clarity and conformance. But then again, I have no prior experience with Powershell and will have it easier to adapt to the "new normal" than others. (I do have preferences that are closer to the way C and Python does things :-))

jborean93 commented on lib/ansible/modules/windows/win_msg.ps1:
```diff
@@ -0,0 +1,65 @@
+$result.sent_localtime = $endsend_at.Trim()
+$result.wait = $wait
+
+If (-not $result.rc -eq 0 ) 
```
If braces again like normal.

jborean93 reviewed (changes requested):
Just some style changes to keep conformity. Only have one question around the -resultobj when getting ansible parameters, what does this do exactly?

Commit:
make bracket style internally consistent

jhawkesworth commented:
Ok @jborean93 - I fixed up the brackets so that they are at least in the same style within this file.

IIRC you used to have to set the -resultobj $results if you used -failifempty true although I'm not sure that is still true any more.

Hope there's enough time to get this in to 2.3.

ready_for_review

jhawkesworth commented:
bot_status

dagwieers commented:
shipit

dagwieers commented:
As mentioned earlier in a discussion, it seems the `-resultobj` is only used for when `Add-AnsibleParam` is bailing out with `Fail-Json`. To be honest, in almost all cases this will be an empty $result because we have not done anything yet. So I would drop this (in python this functionality does not exist either...)

jborean93 commented:
Looks good to me.

shipit

gundalow commented on lib/ansible/modules/windows/win_msg.py:
```diff
@@ -0,0 +1,98 @@
+    default: '*'
+  display_seconds:
+    description:
+      - How long to wait for receiver to acknowledge message.
```
in seconds?

jhawkesworth replied:
Yes, I guess it could be clearer although calling it 'display_seconds' would help.  Do you want me to change the description?

jhawkesworth commented:
I think maybe the idea was so that you could output multiple warnings/errors if there were >1 things wrong with modules params.  Maybe @trondhindenes would recall

Commit:
added units to display_seconds description

jborean93 commented:
In case my shipit was invalidated with the commit.
shipit

dagwieers commented:
shipit

jhawkesworth commented:
Added 'in seconds' to display_seconds parameter description as suggested by @gundalow

jhawkesworth commented:
bot_status

gundalow commented:
Three shipits = merge
Thanks

jhawkesworth commented:
Yay!  Thank you @dagwieers @jborean93 and @gundalow for keeping an eye out for changes on this one today.
